### ISSUE TITLE
The LINQ expression could not be translated (continuous)

### ISSUE THREAD

**Issue by AlexeyMetlikin** [dotnet/efcore#21853]
I have a problem with using OData with EF Core and Automapper.

Guess that there are two tables in DB: Users and Regions

``` C#
    public class Region
    {     
        [Key]
        public virtual Guid Id { get; set; }

        [StringLength(2)]
        [Required]
        public string Code { get; set; }

        [StringLength(100)]
        [Required]
        public string Name { get; set; }

        public virtual User CreatedBy { get; set; }

        [ForeignKey("CreatedBy")]
        public Guid? CreatedById { get; set; }
    }
```

``` C#
    public class User
    {
        [StringLength(50)]
        [Required]
        public string FirstName { get; set; }

        [StringLength(50)]
        [Required]
        public string LastName { get; set; }

        [StringLength(50)]
        public string MiddleName { get; set; }

        public virtual ICollection<Region> CreatedRegions { get; set; }
    }
```

And Also I have DTO models that describe DB Models
``` C#

    public class UserDto
    {
        public Guid Id { get; set; }

        public string FirstName { get; set; }

        public string LastName { get; set; }

        public string MiddleName { get; set; }
    }


    public class RegionDto
    {
        public Guid Id { get; set; }

        public string Code { get; set; }

        public string Name { get; set; }
        
        public UserDto CreatedBy { get; set; }
    }
```

I try to use OData filtering to filter some regions by users who created the regions. For this I have a controller-

``` C#
    [ODataRoutePrefix("[controller]")]
    public class RegionsController : ControllerBase
    {
        private readonly IRegionRepository _regionRepository;

        public RegionsController(IRegionRepository regionRepository)
        {
            _regionRepository = regionRepository;
        }

        [HttpGet]
        [Route("api/[controller]/withRegionFilter")]
        public async Task<List<RegionDto>> GetRegions(ODataQueryOptions<Region> queryOptions)
        {
            return await _regionRepository.GetAll(queryOptions);
        }
        
        [HttpGet]
        [Route("api/[controller]/withRegionDtoFilter")]
        public async Task<List<RegionDto>> GetRegions2(ODataQueryOptions<RegionDto> queryOptions)
        {
            return await _regionRepository.GetAll2(queryOptions);
        }
    }
```

RegionRepository looks like this

``` C#
public interface IRegionRepository
    {
        Task<List<RegionDto>> GetAll(ODataQueryOptions<Region> queryOptions);

        Task<List<RegionDto>> GetAll2(ODataQueryOptions<RegionDto> queryOptions);
    }

    public class RegionRepository : IRegionRepository
    {
        private readonly IMapper _mapper;

        public RegionRepository(IMapper mapper)
        {
            _mapper = mapper;
        }

        public async Task<List<RegionDto>> GetAll(ODataQueryOptions<Region> queryOptions)
        {
            await using var context = new DbContext("ConnectionString");
            return await ((IQueryable<Region>) queryOptions.ApplyTo(context.Set<Region>()))
                .ProjectTo<RegionDto>(Mapper.ConfigurationProvider)
                .ToListAsync();
        }
        
        public async Task<List<RegionDto>> GetAll2(ODataQueryOptions<RegionDto> queryOptions)
        {
            await using var context = new DbContext("ConnectionString");
            return await ((IQueryable<RegionDto>) queryOptions.ApplyTo(context.Set<Region>()
                    .ProjectTo<RegionDto>(Mapper.ConfigurationProvider)))
                .ToListAsync();
        }
    }
```

And, finally, my configuration is 

``` C#
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            app.UseRouting();

            app.UseMvc(routeBuilder =>
            {
                routeBuilder.Filter().Select();
                routeBuilder.EnableDependencyInjection();
                routeBuilder.MapODataServiceRoute("odata", "api", GetEdmModel());
            });
        }

        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers(mvcOptions =>mvcOptions.EnableEndpointRouting = false);
            services.AddOData();
        }

        private static IEdmModel GetEdmModel()
        {
            var odataBuilder = new ODataConventionModelBuilder();

            odataBuilder.EntitySet<RegionDto>("RegionsDto");
            odataBuilder.EntitySet<Region>("Regions");
            odataBuilder.EntitySet<UserDto>("Users");

            return odataBuilder.GetEdmModel();
        }
```

When I try to call api/regions/withRegionFilter?$filter=$it/createdBy/firstName eq 'User', I don't have any problems. An expression tree is translated to this SQL code: 
``` C#
SELECT r."Code", (u."Id" IS NULL), u."FirstName", u."Id", u."LastName", u."MiddleName", r."Id", r."Name"
FROM "Regions" AS r
LEFT JOIN "Users" AS u ON r."CreatedById" = u."Id"
WHERE u."FirstName" = @__TypedProperty_0
```
and it executes perfectly and I have a correct result.

But, when I try to call api/regions/withRegionDtoFilter?$filter=$it/createdBy/firstName eq 'User', I have an exception with message "The LINQ expression could not be translated". In exception I also have an expression tree:
 ``` C#
DbSet<Region>
    .LeftJoin(
        outer: DbSet<User>, 
        inner: r => EF.Property<Nullable<Guid>>(r, "CreatedById"), 
        outerKeySelector: u => EF.Property<Nullable<Guid>>(u, "Id"), 
        innerKeySelector: (o, i) => new TransparentIdentifier<Region, User>(
            Outer = o, 
            Inner = i
        ))
    .Where(r => EF.Property<Nullable<Guid>>(r.Outer.Inner, "Id") == null ? null : new UserDto{ 
        FirstName = r.Outer.Inner.FirstName, 
        Id = r.Outer.Inner.Id, 
        LastName = r.Outer.Inner.LastName, 
        MiddleName = r.Outer.Inner.MiddleName
    }
    .FirstName == __TypedProperty_0)
```

I suppose, it's not correct behavior but I could make a mistake
 
### Further technical details

EF Core version: 3.1.4
Microsoft.AspNetCore.OData: 7.4.1
Database provider: Npgsql.EntityFrameworkCore.PostgreSQL
Target framework: .NET Core 3.1
Operating system: Windows 10 x64
IDE: JetBrains Rider 2019.3.1 Build #RD-193.5662.14, built on December 20, 2019

**Comment by smitpatel** [comment]
Duplicate of #12148

**Comment by AlexeyMetlikin** [comment]
Why was it closed? 
It's duplicate by text error, but error is really different - the expression tree can't be converted to SQL due to it was built wrong. How does it depend on #12148?

**Comment by smitpatel** [comment]
The error is same in both cases. Property access on a conditional expression with either IfTrue/IfFalse being null is failing to translate to SQL.

**Comment by AlexeyMetlikin** [comment]
But it 's worked early, hasn't it?
Is It problem linked with automapper? Due to without automapper it works fine

**Comment by AlexeyMetlikin** [comment]
And can you help me how I can do properties "CreatedBy" and "EditedBy" not nullable?
I tried to do this method with Required attribute and change type from Guid? to Guid, but it doesn't help anyway - everywhere I see    such records as ('r => EF.Property<Nullable<Guid>>(r, "CreatedById")"), f.e.
